### PR TITLE
Improve the marshal module

### PR DESCRIPTION
**Pull request by konsumlamm**
* small changes in the doc comments
* use `runnableExamples`
* refactor the tests
  - delete the tests in `marshal.nim` (they're already in `tmarshal.nim`)
  - use `doAssert` instead of `echo` and `discard`
  - use `block`s
  - swap test 6 & 7

### CODE REVIEW

File: lib/pure/marshal.nim [hunk 1]
```diff
@@ -10,39 +10,33 @@
...
+  assert $$a[] == "{}" # not "{f: 0}"
+
+  # unmarshal
+  let c = to[B]("""{"f": 2}""")
+  assert typeof(c) is B
+  assert c.f == 2
+
+  # marshal
+  let s = $$c
+  assert s == """{"f": 2}"""
```
Comment (timotheecour) [inline lib/pure/marshal.nim]:
```suggestion
  assert $$c == """{"f": 2}"""
```

File: lib/pure/marshal.nim [hunk 2]
```diff
@@ -306,7 +302,7 @@ proc store*[T](s: Stream, data: T) =
 proc `$$`*[T](x: T): string =
   ## Returns a string representation of `x` (serialization, marshalling).
   ##
-  ## **Note:** to serialize `x` to JSON use `$(%x)` from the ``json`` module.
+  ## **Note:** to serialize `x` to JSON use `$(%x)` from the `json` module.
```
Comment (timotheecour) [inline lib/pure/marshal.nim]:
```suggestion
  ## **Note:** to serialize `x` to JSON use `%x` from the `json` module, or `jsonutils. toJson(x)`
```

File: tests/stdlib/tmarshal.nim [hunk 3]
```diff
@@ -148,10 +128,10 @@ block:
...
 
   let test = to[LegacyEntry](str)
-  echo test
+  doAssert $test == """(numeric: "")"""
 
 # bug #16022
 block:
-  let p: proc () = proc () = echo "hello world"
+  let p: proc (): string = proc (): string = "hello world"
   let poc = (to[typeof(p)]($$p))
```
Comment (timotheecour) [inline tests/stdlib/tmarshal.nim]:
```suggestion
  let poc = to[typeof(p)]($$p)
```

File: tests/stdlib/tmarshal.nim [hunk 4]
```diff
@@ -135,9 +116,8 @@ block:
   let
     a1 = some(newJNull())
     a2 = none(JsonNode)
-  echo ($$a1).to[:Option[JsonNode]]
-  echo ($$a2).to[:Option[JsonNode]]
-
+  doAssert $($$a1).to[:Option[JsonNode]] == "Some(null)"
+  doAssert $($$a2).to[:Option[JsonNode]] == "None[JsonNode]"
```
Comment (timotheecour) [inline tests/stdlib/tmarshal.nim:120]:
```suggestion
  doAssert $($$a1).to[:Option[JsonNode]] == "Some(null)"
  doAssert $($$a2).to[:Option[JsonNode]] == "None[JsonNode]"
  doAssert ($$a1).to[:Option[JsonNode]] == some(newJNull())
  doAssert ($$a2).to[:Option[JsonNode]] == none(JsonNode)
```

File: tests/stdlib/tmarshal.nim [hunk 5]
```diff
@@ -60,73 +41,73 @@ proc buildList(): PNode =
 var test3: TestObj
 test3.test = 42
 test3.test2 = blah
```
Comment (timotheecour) [inline tests/stdlib/tmarshal.nim]: test3 = TestObj(test2: blah)
```suggestion
let test3 = TestObj(test2: blah, test: 42)
```

to avoid the `CaseTransition` warning

File: lib/pure/marshal.nim [hunk 6]
```diff
@@ -341,77 +337,3 @@ proc to*[T](data: string): T =
...
-    A {.inheritable.} = object
-    B = object of A
-      f: int
-
-  var
-    a: ref A
-    b: ref B
-  new(b)
-  a = b
-  echo($$a[]) # produces "{}", not "{f: 0}"
```
Comment (timotheecour) [inline lib/pure/marshal.nim:417]: the PR removed the test with {.inheritable.}

**Review by timotheecour — APPROVED** [review]
* LGTM after nits (in particular https://github.com/nim-lang/Nim/pull/16777/files#r561556952)
* followup PR welcome to add static tests (right now there are none!) using the usual pattern:
```nim
template main = # all tests here
static: main()
main()
```

right now this would (after fixing basic issues) give SIGSEGV; an ideal followup PR would identify which of those tests are failing in nimvm, file an issue, and disable just these with the usual:
```nim
when nimvm: discard # xxx see bug #xyz
else: ...
```
* I wonder what benefits does std/marshal have over the newer std/jsonutils, which, unlike marshal:
  * works in js
  * works in VM without SIGSEGV
  * works better with tuples (refs https://github.com/timotheecour/Nim/issues/543)
  * probably works better in other cases too (eg object variants); need to check that point for marshal

**Review by timotheecour — APPROVED** [review]
